### DISCUSSION TITLE
Imputer behavior - test set information leakage problem

### DISCUSSION THREAD

**Discussion by GrzegorzRut** [alan-turing-institute/sktime#2267] [Q&A]
Hi! I just posted a question on WindowSummarizer (https://stackoverflow.com/questions/71561875/forecasting-with-windowsummarizer-and-exogenous-features). When i tried to produce a minimal working example I noticed that Imputer does not seem to remember the values it should use to impute nans - for new datasets it calculates means and medians (maybe other aggregations as well) from scratch. Is it a bug or is it a feature? In standard ML this would not be appropriate.

MInimal working example:
```
from sktime.transformations.series.impute import Imputer
from sktime.datasets import load_longley
_, X = load_longley()

imp = Imputer(method='mean')
imp.fit(X)

nans = X.copy()
nans.loc[:] = np.nan

imp.transform(nans)
```

**Comment by fkiraly** [comment]
@GrzegorzRut, this is an interesting question.

I looked into the imputer, and indeed the mean and median seen in transform is used.

This is *ok* in a time series classification setting, since entire series are either completely in the training set, or the test set.

As you say, however, correctly, this is *not* ok if you do forecasting - the algorithm is not supposed to use any information about the future, this can lead to information leakage and over-optimistic evaluation.

FYI @aiwalter.

I've added leakage free versions of mean/median in #2269, and opened an issue for more complex methods in #2270 (e.g., `forecaster_fit`, together with the naive forecaster this supports a number of common methods).

**Reply by fkiraly** [reply]
also pinging @danbartl as the author of `WindowSummarizer`

**Reply by GrzegorzRut** [reply]
Many thanks for the answer!
